drivers/ethernet/eth_gecko.c

```diff
@@ -0,0 +1,680 @@
+{
+	struct eth_gecko_dev_data *const dev_data = DEV_DATA(dev);
+	const struct eth_gecko_dev_cfg *const cfg = DEV_CFG(dev);
+	ETH_TypeDef *eth = cfg->regs;
+	struct net_pkt *rx_frame = NULL;
+	u16_t frag_len, total_len;
+	u32_t sofIdx, eofIdx;
+	u32_t i, j;
+
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG(dev_data != NULL);
+	__ASSERT_NO_MSG(dev_cfg != NULL);
+
+	/* Preset indeces and total frame length */
+	sofIdx = UINT32_MAX;
+	eofIdx = UINT32_MAX;
+	total_len = 0;
+
+	/* Check if a full frame is received (SOF/EOF present)
+	 *  and determine total length of frame
+	 */
+	for (i = 0; i < ETH_RX_BUF_COUNT; i++) {
+		j = (i + rx_buf_idx);
+		if (j >= ETH_RX_BUF_COUNT) {
+			j -= ETH_RX_BUF_COUNT;
+		}
+
+		/* Verify it is an ETH owned buffer */
+		if (!(dma_rx_desc_tab[j].address & ETH_RX_OWNERSHIP)) {
+			/* No more ETH owned buffers to process */
+			break;
+		}
+
+		/* Check for SOF */
+		if (dma_rx_desc_tab[j].status & ETH_RX_SOF) {
+			sofIdx = j;
+		}
+
+		if (sofIdx != UINT32_MAX) {
+			total_len += (dma_rx_desc_tab[j].status &
+				ETH_RX_LENGTH);
+
+			/* Check for EOF */
+			if (dma_rx_desc_tab[j].status & ETH_RX_EOF) {
+				eofIdx = j;
+				break;
+			}
+		}
+	}
+
+	LOG_DBG("sof/eof: %u/%u, rx_buf_idx: %u, len: %u", sofIdx, eofIdx,
+		rx_buf_idx, total_len);
+
+	/* Verify we found a full frame */
+	if (eofIdx != UINT32_MAX) {
+		/* Allocate room for full frame */
+		rx_frame = net_pkt_rx_alloc_with_buffer(dev_data->iface,
+					total_len, AF_UNSPEC, 0, K_NO_WAIT);
+		if (!rx_frame) {
+			LOG_ERR("Failed to obtain RX buffer");
+			ETH_RX_DISABLE();
+			eth_init_rx_buf_desc();
+			eth->RXQPTR = (u32_t)dma_rx_desc_tab;
+			ETH_RX_ENABLE();
+			return rx_frame;
+		}
+
+		/* Copy frame (fragments)*/
+		j = sofIdx;
+		while (total_len) {
+			frag_len = MIN(total_len, ETH_RX_BUF_SIZE);
+			LOG_DBG("frag: %u, fraglen: %u, rx_buf_idx: %u", j,
+				frag_len, rx_buf_idx);
+			if (net_pkt_write(rx_frame, &dma_rx_buffer[j],
+					  frag_len) < 0) {
+				LOG_ERR("Failed to append RX buffer");
+				dma_rx_desc_tab[j].address &=
+					~ETH_RX_OWNERSHIP;
+				net_pkt_unref(rx_frame);
+				rx_frame = NULL;
+				break;
+			}
+
+			dma_rx_desc_tab[j].address &= ~ETH_RX_OWNERSHIP;
+
+			total_len -= frag_len;
+			if (++j >= ETH_RX_BUF_COUNT) {
+				j -= ETH_RX_BUF_COUNT;
+			}
+
+			if (++rx_buf_idx >= ETH_RX_BUF_COUNT) {
+				rx_buf_idx -= ETH_RX_BUF_COUNT;
+			}
+		}
+	}
+
+	return rx_frame;
+}
+
+static void eth_rx(struct device *dev)
+{
+	struct eth_gecko_dev_data *const dev_data = DEV_DATA(dev);
+	struct net_pkt *rx_frame;
+	int res = 0;
+
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG(dev_data != NULL);
+
+	/* Iterate across (possibly multiple) frames */
+	rx_frame = frame_get(dev);
+	while (rx_frame) {
+		/* All data for this frame received */
+		res = net_recv_data(dev_data->iface, rx_frame);
+		if (res < 0) {
+			LOG_ERR("Failed to enqueue frame into RX queue: %d",
+			res);
+			/*eth_stats_update_errors_rx(dev_data->iface);*/
```

Comment:
You mentioned in the beginning of the file that statistics are not supported. To me it looks quite trivial task to support that as you only need the interface pointer for it.

Reply:
I actually hadn't taken a good look yet at what supporting statistics would require. I uncommented line 232 in the section above, is that enough for this driver or are there other things I should support? If so, could you point me to a driver that I can use as a starting point?

Reply:
It seems that currently all the existing drivers just have `eth_stats_update_errors_rx(context->iface);` to update the RX error count. So here you could just uncomment this line and it should work just fine. Just add `#include <ethernet/eth_stats.h>` at the beginning of the file. Currently all the other Ethernet stats are updated by the generic Ethernet L2 driver.

Reply:
I see, thank you! I've uncommented `eth_stats_update_errors_rx(context->iface);` and included `<ethernet/eth_stats.h>`.